### PR TITLE
refactor: Improve logger interface, use logger attributes by <placeholder>

### PR DESCRIPTION
**Pull request by michaljurecko**
Jira: https://keboola.atlassian.net/browse/PSGO-446

**Changes:**
- Improed logger interface, `Log` method has been splitted to `Log` and `Logf`.
- Added support for `Logger.With` attributes in placeholders.

-----------

### CODE REVIEW

File: internal/pkg/service/stream/storage/cleanup/node.go [hunk 1]
```diff
@@ -111,7 +111,7 @@ func (n *Node) cleanFiles(ctx context.Context) (err error) {
 	defer func() {
 		cnt := count.Load()
 		span.SetAttributes(attribute.Int64("deletedFilesCount", cnt))
-		n.logger.With(attribute.Int64("count", cnt)).Infof(ctx, `deleted "%d" files`, cnt)
+		n.logger.With(attribute.Int64("deletedFilesCount", cnt)).Info(ctx, `deleted "<deletedFilesCount>" files`)
 	}()
```
Comment (michaljurecko) [inline internal/pkg/service/stream/storage/cleanup/node.go:115]: We want to use `Logger.With` attributes in placeholders, it was discussed multiple times.

https://github.com/keboola/keboola-as-code/pull/1600#discussion_r1497336806

File: internal/pkg/log/memory_test.go [hunk 2]
```diff
@@ -16,24 +18,25 @@ func TestMemoryLogger(t *testing.T) {
...
 		With(attribute.String("key1", "value1"), attribute.String("key2", "value2")).
 		WithComponent("c2").
 		With(attribute.String("key3", "value3")).
 		WithDuration(123 * time.Second)
 
-	memWithCtx.Debug(context.Background(), `Debug message.`)
-	memWithCtx.Info(context.Background(), `Info message.`)
+	ctx := ctxattr.ContextWith(context.Background(), attribute.String("key4", "value4"))
+	memWithAttrs.Debug(ctx, `Debug message - <key1> <key2> <key3> <key4>`)
+	memWithAttrs.Info(ctx, `Info message - <key1> <key2> <key3> <key4>`)
```
Comment (michaljurecko) [inline internal/pkg/log/memory_test.go:30]: So we need to make `Logger.With` attributes equal to `ctxattrs`.

File: internal/pkg/log/log.go [hunk 3]
```diff
@@ -46,21 +46,27 @@ type DebugLogger interface {
...
+	// Infof logs formatted message in the info level, you can use an attribute <placeholder> for ctxattr or Logger.With attributes.
 	Infof(ctx context.Context, template string, args ...any)
+	// Warnf logs formatted message in the warning level, you can use an attribute <placeholder> for ctxattr or Logger.With attributes.
 	Warnf(ctx context.Context, template string, args ...any)
+	// Errorf logs formatted message in the error level, you can use an attribute <placeholder> for ctxattr or Logger.With attributes.
 	Errorf(ctx context.Context, template string, args ...any)
-
-	Log(ctx context.Context, level string, args ...any)
+	// Logf logs formatted message in the level, you can use an attribute <placeholder> for ctxattr or Logger.With attributes.
+	Logf(ctx context.Context, level string, template string, args ...any)
```
Comment (michaljurecko) [inline internal/pkg/log/log.go:70]: I fixed also the `Logger` interface: the inconsistency in the `Log` method, now there are two methods, `Log` and `Logf`.

File: internal/pkg/log/logger.go [hunk 4]
```diff
@@ -19,22 +19,28 @@ const ComponentKey = "component"
 // zapLogger is default implementation of the Logger interface.
 // It is wrapped zap.SugaredLogger.
 type zapLogger struct {
-	sugaredLogger *zap.SugaredLogger
-	core          zapcore.Core
-	component     string
+	logger     *zap.Logger
+	core       zapcore.Core
+	attributes []attribute.KeyValue
```
Comment (michaljurecko) [inline internal/pkg/log/logger.go:24]: There is no `core.Attributes` getter, so I added the `attributes` field.

File: internal/pkg/log/logger.go [hunk 5]
```diff
@@ -52,114 +58,94 @@ func (l *zapLogger) WithDuration(v time.Duration) Logger {
...
+
+func (l *zapLogger) message(ctx context.Context, message string) string {
+	var replacements []string
+	for _, kv := range ctxattr.Attributes(ctx).ToSlice() {
+		replacements = append(replacements, "<"+string(kv.Key)+">", kv.Value.Emit())
+	}
+	for _, kv := range l.attributes {
+		replacements = append(replacements, "<"+string(kv.Key)+">", kv.Value.Emit())
+	}
+	return strings.NewReplacer(replacements...).Replace(message)
```
Comment (michaljurecko) [inline internal/pkg/log/logger.go:138]: The both, `ctxattrs` and `logger.attributes` are used when formatting the message.

**Review by hosekpeter — APPROVED** [review]

**Pull request merged**